File: Android/app/src/main/java/com/hhp227/datemate/ui/main/lounge/LoungeViewModel.kt
package com.hhp227.datemate.ui.main.lounge

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.datemate.common.Resource
import com.hhp227.datemate.data.PostRepository
import com.hhp227.datemate.data.Post
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import java.lang.Exception

class LoungeViewModel(
    private val repository: PostRepository
) : ViewModel() {
    val state = mutableStateOf(State())

    private fun getPosts() {
        repository.getPosts().map(::getPostsUseCase).flowOn(Dispatchers.IO).onEach(::onReceive).launchIn(viewModelScope)
    }

    private fun getPostsUseCase(batch: List<Post>): Resource<List<Post>> {
        return try {
            Resource.Success(batch)
        } catch (e: Exception) {
            Resource.Error(e.localizedMessage ?: "An unexpected error occured")
        }
    }

    private fun onReceive(result: Resource<List<Post>>) {
        when (result) {
            is Resource.Success -> {
                this.state.value = State(posts = result.data ?: emptyList())
            }
            is Resource.Error -> {
                this.state.value = State(error = result.message ?: "An unexpected error occured")
            }
            is Resource.Loading -> {
                this.state.value = State(isLoading = true)
            }
        }
    }

    init {
        getPosts()
    }

    data class State(var isLoading: Boolean = false, var posts: List<Post> = emptyList(), var error: String = "")
}

File: iOS/DateMate/UI/Main/Lounge/LoungeViewModel.swift
//
//  LoungeViewModel.swift
//  DateMate
//
//  Created by 홍희표 on 2021/12/07.
//

import Foundation
import Combine

class LoungeViewModel: ObservableObject {
    @Published var state = State()
    
    //@Published var uiEvent = PassthroughSubject<UiEvent, Error>()
    
    private let repository: LoungeRepository
    
    private var subscription = Set<AnyCancellable>()
    
    private func getPosts() {
        /*repository.getPosts().tryMap(getPostsUseCase).receive(on: DispatchQueue.main).sink(receiveCompletion: onReceive, receiveValue: onReceive).store(in: &subscription)*/
    }
    
    private func getPostsUseCase(posts: [Post]) -> Resource<[Post]> {
        do {
            return Resource.success(data: posts)
        } catch {
            return Resource.error(message: error.localizedDescription)
        }
    }
    
    private func onReceive(_ result: Resource<[Post]>) {
        switch result.state {
        case .Success:
            self.state = State(posts: result.data ?? [Post]())
        case .Error:
            self.state = State(error: result.message ?? "An unexpected error occured")
        case .Loading:
            self.state = State(isLoading: true)
        }
    }
    
    func onReceive(_ completion: Subscribers.Completion<Error>) {
        switch completion {
        case .finished:
            break
        case .failure:
            break
        }
    }
    
    func test() {
    }
    
    /*func onEvent(_ event: LoungeEvent) {
        switch event {
        case .OnPostClick:
            sendUiEvent(.Navigate(string: "Route_ADD_EDIT_POST + PostId"))
        case .OnAddPostClick:
            sendUiEvent(.Navigate(string: "Route_ADD_EDIT_POST"))
        }
    }
    
    private func sendUiEvent(_ event: UiEvent) {
        DispatchQueue.main.async {
            self.uiEvent.send(event)
        }
    }*/
    
    init(_ repository: LoungeRepository) {
        self.repository = repository
        
        getPosts()
    }
    
    deinit {
        subscription.removeAll()
    }
    
    struct State {
        var isLoading = false
        
        var posts = [Post]()
        
        var error = ""
        
    }
}

enum LoungeEvent {
    case OnPostClick(post: Post), OnAddPostClick
}
